### ISSUE TITLE
Idea Hub draft post ideas requested on every page load

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3757]
## Bug Description

This is a follow-up bug for #3733.

See https://github.com/google/site-kit-wp/issues/3733#issuecomment-886938896: Currently the request to get draft post ideas in the dashboard happens on every page load, which is undesirable for performance.

In addition, the `getNewIdeas` and `getSavedIdeas` selectors also consider draft posts for ideas (in order to exclude those from the responses), but the original fix doesn't address that. Essentially, whatever mechanism to invalidate the cache based on Idea Hub post changes `getDraftPostIdeas` gets, we also need to do for `getNewIdeas` and `getSavedIdeas`.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Whenever an Idea Hub-based post has its post status changed or is deleted, that time should be stored in a transient for an hour, essentially as a "when was any Idea Hub post last changed" value.
* On the client-side, Idea Hub caches should be invalidated for any already cached results that are older than that stored timestamp (if there is any). This should apply to any Idea Hub endpoint caches to get ideas.
* The server-side timestamp should be in a transient with one hour expiration, since it wouldn't be meaningful after more than an hour anyway, as the client-side cache expiration is also one hour.

### Generate last changed time for Idea Hub posts

In `includes/Modules/Idea_Hub.php`:

* Add a new `const IDEA_HUB_LAST_IDEA_POST_UPDATED_AT = googlesitekit_idea_hub_last_idea_post_updated_at`
* In `$this->is_connected() )` function, add a hook for `transition_post_status` action (https://codex.wordpress.org/Post_Status_Transitions )
  * Return if the post type is not an idea hub post (see `is_idea_post()`)
  * Update the transient `IDEA_HUB_LAST_IDEA_POST_UPDATED_AT` with the output of `time()` using the `set` method of the `Transient` class instance with `$expiration` param set to `0`.
* Add a new instance of the `Script_Data` class to the Idea Hub assets returned from the `setup_assets` method. It should use the `googlesitekit-idea-hub-data` handler and the following options passed as the second argument to its constructor:
  * The `global` property should be `_googlesitekitIdeaHub`
  * The `data_callback` property should be a function that returns an array with the `lastIdeaPostUpdatedAt` property.

### <del>Create an endpoint for last updated param for Idea Hub posts</del>

 * <del>Create a new endpoint in `get_datapoint_definitions()` and `create_data_request()` for `GET:last-changed-timestamp` to return the `IDEA_HUB_LAST_CHANGED` transient value.</del>
 * <del>In `$this->is_connected() )` function, add a hook for `googlesitekit_apifetch_preload_paths` filter to preload the url. See the following for example: https://github.com/google/site-kit-wp/blob/bbc5d84de6869937294a5009a9c783c4ca5ae8d5/includes/Core/Feature_Tours/REST_Feature_Tours_Controller.php#L62-L71</del>

### Add <del>a new fetch datastore and</del> a new selector to the Idea Hub module that will return the timestamp

 * <del>Create a new data store via `createFetchStore` in `assets/js/modules/idea-hub/datastore/last-changed-timestamp.js`</del>
   * <del>`const fetchGetLastChangedTimestamp`</del>
   * <del>`basename getLastChangedTimestamp`
   * <del>return API.get( 'modules', 'idea-hub', 'last-changed-timestamp', undefined, { useCache: false, } );</del>
   * <del>In `baseInitialState` add `lastChangedTimestamp: undefined`</del>
   * <del>In `baseResolvers` and `baseSelectors` add `getLastChangedTimestamp`</del>
   * <del>Export the store.</del>
* Add a new action `receiveIdeaHubData` that receives an object with custom data for the Idea Hub module and returns an action with `RECEIVE_IDEA_HUB_DATA` type and the incoming data as the action payload object.
* Add a new case to the reducer for the new action that will add Idea Hub data to the state.
* Add a new resolver `getLastIdeaPostUpdatedAt` that calls the `receiveIdeaHubData` action and passes the `_googlesitekitIdeaHub` object into it.
* Add a new selector `getLastIdeaPostUpdatedAt` that returns the `lastIdeaPostUpdatedAt` value from the Idea Hub data state.

_See how site info datastore works. It uses the same approach to read global variables and add them to the store state._

### Update exsiting datastores to use last changed timestamp

Update the following three places in a similar manner:
* `assets/js/modules/idea-hub/datastore/draft-post-ideas.js`
* `assets/js/modules/idea-hub/datastore/new-ideas.js`
* `assets/js/modules/idea-hub/datastore/saved-ideas.js`

```diff
 // new-ideas.js

 const fetchGetNewIdeasStore = createFetchStore( {
 	baseName: 'getNewIdeas',
-	controlCallback: () => {
- 		return API.get( 'modules', 'idea-hub', 'new-ideas' );
+	controlCallback: ( { timestamp } ) => {
+		return API.get( 'modules', 'idea-hub', 'new-ideas', { timestamp } );
 	},
 	reducerCallback: ( state, newIdeas ) => {
 		return {
 			...state,
 			newIdeas,
 		};
 	},
+	argsToParams( { timestamp } ) {
+		return { timestamp };
+	},
 } );

 ...

 const baseResolvers = {
 	*getNewIdeas() {
 		const registry = yield commonActions.getRegistry();
 		const newIdeas = registry.select( MODULES_IDEA_HUB ).getNewIdeas();
 
 		// If there are already ideas in state, don't make an API request.
 		if ( newIdeas === undefined ) {
-			yield fetchGetNewIdeasStore.actions.fetchGetNewIdeas();
+			const timestamp = registry.select( MODULES_IDEA_HUB ).getLastIdeaPostUpdatedAt();
+			yield fetchGetNewIdeasStore.actions.fetchGetNewIdeas( { timestamp } );
 		}
 	},
 };

 ...
```

### Test Coverage

* <del>Create a new test file in `assets/js/modules/idea-hub/datastore/last-changed-timestamp.test.js`</del>

* Update the following tests to handle cached data:
  * `assets/js/modules/idea-hub/datastore/new-ideas.test.js`
  * `assets/js/modules/idea-hub/datastore/saved-ideas.test.js`
  * `assets/js/modules/idea-hub/datastore/draft-post-ideas.test.js`

### Visual Regression Changes

* N/A

### Overview
This shouldn't be `QA:eng` as it is user-facing, although you will need to do some slightly technical checks (looking at network requests) to completely verify the ACs have been met.

### Set up
* You will need Idea Hub enabled and connected.
* Open two browser windows: 
  * one with the Site Kit dashboard
  * the other showing the WordPress Posts page (`/wp-admin/edit.php`)
* On the tab showing the dashboard, open Chrome dev tools and select the `Network` tab. Select `XHR` along the top and using the 'filter' input, type 'idea' to filter for idea hub requests (see image below).
  
![Screenshot 2021-08-12 at 10 40 22](https://user-images.githubusercontent.com/2470911/129175815-4568cae7-0d17-4905-86a4-ea120fae885b.png)

### Requests are made to the endpoints on page load.
- Delete the application cache by going to the `Application` tab in dev tools and using the 'clear all' (🚫 ) button. 
- Switch back to the network tab (see setup instructions and image above).
- Then refresh the browser. 
- There should be 3 new Idea Hub requests to three different endpoints:`new-ideas`, `saved-ideas` and `draft-post-ideas`. 

### Subsequent page loads should then use the cache (no new requests).
- Fresh requests should only be made when something has changed, so refresh the Site Kit dashboard **without** interacting with the Idea Hub widget.
- There should be no new Idea Hub requests in the network tab (because the cache is being used).

### Creating a draft idea post should invalidate the cache.
- Create a draft using the 'create a draft' button. 
- The number in the 'Drafts' tab should increment by one. 
- This should make a fresh request to the `create-idea-draft-post` endpoint and then wipe the cache.
- This means that if you refresh again, you should expect to see fresh requests made to the `new-ideas`, `saved-ideas` and `draft-post-ideas` endpoints.

### Changing an Idea Hub post state should force fresh requests
- Having created a draft post using the widget, now switch to the other browser window (showing your WordPress posts) and refresh to see the draft post appear.
- Change the state of the post (for example, trash it).
- Now switch back to the browser window showing the Site Kit dashboard and refresh
- Fresh requests should have been made to all 3 endpoints, and the number in the 'Drafts' tab should be adjusted accordingly (e.g. if you trashed the post, it should decrement).

### Changing a non-Idea Hub post state should not force fresh requests
- Switch back to the browser window showing the posts page.
- Change the state of a *non*-Idea Hub post.
- As before, switch back to the dashboard and refresh.
- No new requests should have been made.

## Changelog entry

* Update Idea Hub new, saved, and draft ideas requests to invalidate the cache when a post created for an idea changes its status.

**Comment by felixarntz** [comment]
@tofumatt @johnPhillips I've opened this one to follow up on #3733 - let's continue discussing potential approaches here.

**Comment by tofumatt** [comment]
My concern with the original approach of clearing the client-side cache is that we need to implement a whole mechanism for that—we need a way to set something on the client before any components mount that clears the cache… and clearing the cache is async.

I guess the best option would be to put this in a very high `useMount` or something that happens before we even do a `React.render`.

It's totally possible, but it's just a lot more complicated. But if we're doing this it might be nice to make a generic way to do this?

@ivankruchkoff mentioned on a call he is already working through this… so many all of this is already covered, but just documenting my thoughts here 🙂

**Comment by felixarntz** [comment]
@tofumatt Makes sense, we could have it in a new component that is included at a high level, e.g. via `Root`.

Regarding a generic way, it would be great to find one here, but we also shouldn't make it _too_ complicated. I think the approach of only supporting the Idea Hub-specific case for now is reasonable, since we have no idea what other conditions could come up later that would require certain caches to be cleared. But we could surely have an abstraction where we e.g. allow passing a map of JS flag names for certain events that come from the server-side, along with all the caches that should be invalidated if they are older than the timestamp in the flag.

**Comment by eugene-manuilov** [comment]
@ivankruchkoff thanks for IB, but I would agree with Felix that we don't need to use an over-engineered approach. Let's just support only the Idea Hub-specific case for now. I think what we can do for now is the following:

1. Add a hook that will save the timestamp of the latest idea post status change (we don't even need to cache it for an hour, it can be non-expirable).
1. Add a new endpoint to the Idea Hub module that will return the timestamp
1. Add this new endpoint to the API preload paths list
1. Add a new fetch datastore and a new selector to the Idea Hub module that will return the timestamp
1. Update fetch datastores for new and saved ideas endpoints to add the timestamp to the request query parameters

The additional `timestamp` parameter in the API URL will control API cache invalidation and in the same time will be ignored on the server-side.

Here is a small snippet of how we can change the existing fetch datastores:

```diff
 // new-ideas.js

 const fetchGetNewIdeasStore = createFetchStore( {
 	baseName: 'getNewIdeas',
-	controlCallback: () => {
- 		return API.get( 'modules', 'idea-hub', 'new-ideas' );
+	controlCallback: ( { timestamp } ) => {
+		return API.get( 'modules', 'idea-hub', 'new-ideas', { timestamp } );
 	},
 	reducerCallback: ( state, newIdeas ) => {
 		return {
 			...state,
 			newIdeas,
 		};
 	},
+	argsToParams( { timestamp } ) {
+		return { timestamp };
+	},
 } );

 ...

 const baseResolvers = {
 	*getNewIdeas() {
 		const registry = yield commonActions.getRegistry();
 		const newIdeas = registry.select( MODULES_IDEA_HUB ).getNewIdeas();
 
 		// If there are already ideas in state, don't make an API request.
 		if ( newIdeas === undefined ) {
-			yield fetchGetNewIdeasStore.actions.fetchGetNewIdeas();
+			const timestamp = yield commonActions.await( registry.__experimentalResolveSelect( MODULES_IDEA_HUB ).getLastestIdeaPostTimestamp() );
+			yield fetchGetNewIdeasStore.actions.fetchGetNewIdeas( { timestamp } );
 		}
 	},
 };

 ...
```

cc @felixarntz

**Comment by eugene-manuilov** [comment]
> Update the option IDEA_HUB_LAST_CHANGED with the output of time() via update_option() with $autoload param set to false.

@ivankruchkoff let's better use Transient with `0` expiration time which means no expiration. Everything else seems to be good to me.

**Comment by eugene-manuilov** [comment]
> Update the transient IDEA_HUB_LAST_CHANGED with the output of time() via set_transient() with $expiration param set to 0.

By `Transient` I meant an instance of the `Transient` class as we do it for the admin notices. I'll update it.

**Comment by eugene-manuilov** [comment]
IB :heavy_check_mark: for me. Assigning it to @felixarntz for review as well.

**Comment by felixarntz** [comment]
IB ✅ 

@ivankruchkoff @eugene-manuilov My only thoughts are whether we really need a dedicated API endpoint and all the datastore infrastructure for this single piece of information. I think it would be a lot less work and more straightforward to just expose this in PHP-to-JS data - I know that would not be quite as elegant, but it's also IMO not worth building out a full API infrastructure for it. Thoughts on this?

**Comment by eugene-manuilov** [comment]
> My only thoughts are whether we really need a dedicated API endpoint and all the datastore infrastructure for this single piece of information. I think it would be a lot less work and more straightforward to just expose this in PHP-to-JS data - I know that would not be quite as elegant, but it's also IMO not worth building out a full API infrastructure for it. Thoughts on this?

@felixarntz, yes, PHP-to-JS is a simpler and more straightforward approach, but this is inconsistent with other things that we have in our codebase. We use the API/prefetch/datastore approach for a number of things that can be done much easier using the PHP-to-JS approach.

For example, user tracking status or dismissed feature tours/items - we have API endpoints and datastores for them even though we don't re-fetch data again on the client-side because it is always prefetched (which is technically saying the same PHP-to-JS approach).

I think we should stick with the approach we use across our codebase even though it's a little bit more work. cc @aaemnnosttv

**Comment by aaemnnosttv** [comment]
@eugene-manuilov @felixarntz I think we should still use the datastore, but we can probably do without the fetch-related infrastructure for now and keep it to a simple selector that either resolves from a global variable or sets the initial value from the global with no resolver. There could still be a corresponding action to receive the value for tests which is a convention I think we want to keep going forward. This would be similar to what we do with `core/user` data (e.g. user-info).

The preloaded data for API fetch is a nice middle ground but is more relevant for data that we want to ensure is up-to-date on the server as well as something that would be immediately requested since the preloaded data has a very short TTL. In this case, it sounds like we're not as concerned about this being stale if the value had changed after the page was loaded, in which case fetch infra seems a bit unnecessary.

**Comment by felixarntz** [comment]
@eugene-manuilov The difference between the features you're mentioning for reference and what we're building here IMO is that the one here is very specific to a single module at this point. For example, the dismissing feature tours or notifications aspect is something we wanted to make reusable since it's relevant for multiple areas and is already being used as such. What we have here though, we're building specifically for Idea Hub, so I tend to think a REST endpoint for just this one field is overarchitected, especially since it's not intended to be extended either (see the data point name used, which limits us to _only_ that one timestamp field). Since we're not working towards reusability or extensibility here, I think the PHP-to-JS approach would be more suitable here. I'm not strongly opposed to building the endpoint, but I think we could save some unnecessary time investment by not doing so.

I agree with @aaemnnosttv though that, even in that case, we should still use the data store to access that PHP-to-JS data.

**Comment by eugene-manuilov** [comment]
@felixarntz @aaemnnosttv ok, I have updated IB. Does it look good to you now? cc @johnPhillips

**Comment by aaemnnosttv** [comment]
Thanks @eugene-manuilov – I think that looks good for the most part with a few things to clarify:

- `googlesitekit_idea_hub_last_changed`, `_googlesitekitIdeaHub.lastChangedTimestamp` and the `getLastChangedTimestamp` are a little too generically named - we should be more specific with these names to clarify this is about the last idea hub post updated timestamp. I would suggest something like `googlesitekit_idea_hub_last_idea_post_updated_at` and `getLastIdeaPostUpdatedAt` respectively.

> Add a new selector `getLastChangedTimestamp` that returns the `lastChangedTimestamp` value from the global `_googlesitekitIdeaHub` variable.

Just to clarify that the selectors should still always read the value from `state`. The global should only be used for providing the `initialValue` (or a resolver to set it). As mentioned above, we still want to include a `receive*` action for setting this (mostly for tests, but this would be used by the resolver as well if we go that route).

As for the added `timestamp` param to the fetch actions, I would expect this to be added by the resolver or did you have something else in mind?

**Comment by eugene-manuilov** [comment]
> googlesitekit_idea_hub_last_changed, _googlesitekitIdeaHub.lastChangedTimestamp and the getLastChangedTimestamp are a little too generically named - we should be more specific with these names to clarify this is about the last idea hub post updated timestamp. I would suggest something like googlesitekit_idea_hub_last_idea_post_updated_at and getLastIdeaPostUpdatedAt respectively.

Thanks, @aaemnnosttv. Updated.

> Just to clarify that the selectors should still always read the value from state. The global should only be used for providing the initialValue (or a resolver to set it). As mentioned above, we still want to include a receive* action for setting this (mostly for tests, but this would be used by the resolver as well if we go that route).

Updated this as well. Added new action, reducer, and resolver to IB to facilitate data transfer from the global variable to the store state. Now, this approach is neither simpler nor more straightforward than using an API endpoint and prefetch functionality. :man_shrugging: 

> As for the added timestamp param to the fetch actions, I would expect this to be added by the resolver or did you have something else in mind?

Yes, we add it in resolvers for new/saved/draft ideas selectors. See the snippet in the IB that shows how to use it. Or did you mean something else? :thinking:

**Comment by aaemnnosttv** [comment]
> Updated this as well. Added new action, reducer, and resolver to IB to facilitate data transfer from the global variable to the store state. Now, this approach is neither simpler nor more straightforward than using an API endpoint and prefetch functionality.

The original concern was about creating a REST endpoint for a single bit of data which we can easily avoid. For the sake of time, let's do the simplest thing for now and skip all the datastore stuff and just read the global in the control where we make the request to use in the `timestamp`. We can refactor it later to use the datastore if and when we need it to be changeable or selected elsewhere on the client. cc: @felixarntz 

> Yes, we add it in resolvers for new/saved/draft ideas selectors. See the snippet in the IB that shows how to use it.

Sorry, I didn't see that part because it was hidden by the scroll 😄 Considering the above, it isn't relevant anymore anyways.

**Comment by ivankruchkoff** [comment]
QA: ❌

Timestamp param is added via `getLastIdeaPostUpdatedAt()` selector.

* Disabled Site Kit by Google – Test Idea Hub connection
* Reset sitekit
* Enabled Sitekit
* Enabled idea hub module

First load:
googlesitekit.data.stores["modules/idea-hub"].selectors.getLastIdeaPostUpdatedAt()
false - should this be the time when the module is enabled?

* Enabled Site Kit by Google – Test Idea Hub connection
* Loaded with dummy new posts
* Create draft, then reload dashboard page

googlesitekit.data.stores["modules/idea-hub"].selectors.getLastIdeaPostUpdatedAt()
false

* Trash draft post
* load dashboard again:
googlesitekit.data.stores["modules/idea-hub"].selectors.getLastIdeaPostUpdatedAt()
'1629363401'

* Create draft, then reload dashboard page
googlesitekit.data.stores["modules/idea-hub"].selectors.getLastIdeaPostUpdatedAt()
'1629363401'

Suggestions for improvements, 
1) Update getLastIdeaPostUpdatedAt when module is enabled.
2) When we create a post, due to https://github.com/google/site-kit-wp/pull/3826/files#diff-1963ff009619f8ef5c8ec1ecb2c6b3aad3c93583428f7e6d345c727de15f2204R817 as `$this->is_idea_post( $post->ID )` is always false for a new post creation, since first we create the post, then we add the meta fields to indicate it is an idea post. We can work around this by **also** updating the transient in the `set_post_idea` function:
https://github.com/google/site-kit-wp/blob/f46f548d709f4224a7af75f40701b3b05c9c848f/includes/Modules/Idea_Hub.php#L744

or alternatively in `create-idea-draft-post` handler after adding the idea.

 https://github.com/google/site-kit-wp/blob/f46f548d709f4224a7af75f40701b3b05c9c848f/includes/Modules/Idea_Hub.php#L487

What do you think @eugene-manuilov ?

**Comment by eugene-manuilov** [comment]
> Update getLastIdeaPostUpdatedAt when module is enabled.

@ivankruchkoff The particular value of the `lastIdeaPostUpdatedAt` property doesn't really matter for us. More important for us is the fact that this property value gets changed when a post changes its status. So, I think we don't need to update it when the Idea Hub module is enabled.

> When we create a post, due to https://github.com/google/site-kit-wp/pull/3826/files#diff-1963ff009619f8ef5c8ec1ecb2c6b3aad3c93583428f7e6d345c727de15f2204R817 as $this->is_idea_post( $post->ID ) is always false for a new post creation, since first we create the post, then we add the meta fields to indicate it is an idea post. We can work around this by also updating the transient in the set_post_idea function:

When we create a post, we manually invalidate the cache in the `controlCallback` function which means that we will get the correct results from the server regardless of the `lastIdeaPostUpdatedAt` property value.

https://github.com/google/site-kit-wp/blob/5add4f1d69e0b56aa96d942976a894260029b064/assets/js/modules/idea-hub/datastore/draft-ideas.js#L47-L51

So, to summarize here: we don't care about the value of the `lastIdeaPostUpdatedAt` property. We need to make sure that the draft ideas list gets correctly updated when a post is trashed or restored from the trash.

cc @aaemnnosttv

**Comment by ivankruchkoff** [comment]
Update: QA: ✅
@eugene-manuilov yes, I can confirm that we're invalidating the cache for the three endpoints.

If we don't want to change the functionality, perhaps we should change `lastIdeaPostUpdatedAt` to better reflect that it doesn't account for new draft posts being created, and also that it's initialised with the value `false` rather than when the idea hub module is enabled.
